virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,227 @@
+    .owner = THIS_MODULE,
+    .read = eeprom_read,
+    .write = eeprom_write,
+};
+
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    mm_segment_t old_fs;
+    ssize_t ret;
+    char file_path[128];
+
+    snprintf(file_path, sizeof(file_path), "/mnt/settings/%s.bin", eeprom->name);
+
+    old_fs = get_fs();
+    set_fs(KERNEL_DS);
+    file = filp_open(file_path, O_WRONLY | O_CREAT, 0644);
+    if (IS_ERR(file)) {
+        set_fs(old_fs);
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_write(file, eeprom->data, eeprom->size, &file->f_pos);
+    filp_close(file, NULL);
+    set_fs(old_fs);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    mm_segment_t old_fs;
+    ssize_t ret;
+    char file_path[128];
+
+    snprintf(file_path, sizeof(file_path), "/mnt/settings/%s.bin", eeprom->name);
+
+    old_fs = get_fs();
+    set_fs(KERNEL_DS);
+    file = filp_open(file_path, O_RDONLY, 0644);
+    if (IS_ERR(file)) {
+        set_fs(old_fs);
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_read(file, eeprom->data, eeprom->size, &file->f_pos);
+    filp_close(file, NULL);
+    set_fs(old_fs);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int __init eeprom_init(void)
+{
+    struct device_node *np, *child;
+    int ret;
+    char proc_name[32];
+
+    // Create the /proc/virtual_eeprom directory
+    eeprom_dir = proc_mkdir("virtual_eeprom", NULL);
+    if (!eeprom_dir) {
+        pr_err("Failed to create /proc/virtual_eeprom directory\n");
+        return -ENOMEM;
+    }
+
+    // Find the parent device node
+    np = of_find_node_by_name(NULL, "virtual_eeproms");
+    if (!np) {
+        pr_err("No device node found for virtual_eeproms\n");
+        return -ENODEV;
+    }
+
+    // Iterate through all child nodes
+    for_each_child_of_node(np, child) {
+        const char *name = child->name;
+        u32 size;
+
+        if (of_property_read_u32(child, "size", &size)) {
+            pr_err("Failed to read size for %s\n", name);
+            continue;
+        }
+
+        struct virtual_eeprom *eeprom = kzalloc(sizeof(struct virtual_eeprom), GFP_KERNEL);
+        if (!eeprom) {
+            pr_err("Failed to allocate memory for %s\n", name);
+            continue;
+        }
+
+        eeprom->data = kzalloc(size, GFP_KERNEL);
+        if (!eeprom->data) {
+            pr_err("Failed to allocate memory for EEPROM %s\n", name);
+            kfree(eeprom);
+            continue;
+        }
+
+        mutex_init(&eeprom->mutex);
+        eeprom->size = size;
+        eeprom->name = name;
+
+        ret = load_eeprom_from_file(eeprom);
+        if (ret < 0) {
+            memset(eeprom->data, 0, size);
+        }
+
+        snprintf(proc_name, sizeof(proc_name), "%s", name);
+        eeprom->proc_entry = proc_create_data(proc_name, 0666, eeprom_dir, &eeprom_fops, eeprom);
```

Comment:
A little unorthodox, but I guess if it works...
Normally, you'd create a device node in `/sys` for each device, and then add attributes to it. For instance, in this case, you'd declare a BIN_ATTR and then create it with `sysfs_create_bin_file()`, or something along those lines

Reply:
Yeah what was the reason again that you used procfs instead of sysfs which is generally where device drivers would go?

Reply:
It's easier to add stuff to `procfs` because there's fewer rules, so it has become a bit of a dumping ground. They introduced `sysfs` for this very reason, but I don't quite see how adding something with more rules and complications is going to make people do it the right (hard) way instead of the wrong (easy) way without strict enforcement

Reply:
I did start with `sysfs` implementation, but then I found out that when I do `cat /sys/kernel/virtual_eeprom/eeprom`, `sysfs` calls the `show_data` function for us, and the maximum size of the buffer it allocates is 4096 bytes. procfs implementation does not have this limitation and works good in this case.

Reply:
Not sure what `show_data()` function you are referring to. I'm going to approve this because I don't want to be the gate keeper, but I do think when it comes to kernel modules (and this holds even more true if we're going to show this to a customer) we should stick to best practice - and putting attributes for a device in procfs isn't best practice.
Leveraging the nvmem subsystem would be a bonus: https://github.com/torvalds/linux/blob/bc83b4d1f08695e85e85d36f7b803da58010161d/Documentation/devicetree/bindings/nvmem/nvmem.yaml